File: ios/Runner/AppUsageImplementation.swift
//
//  AppUsageImplementation.swift
//  Runner
//
//  Created by Pranav Masekar on 29/09/25.
//

import Foundation

class AppUsageImplementation: AppUsageApi {
    
    func getPlatformVersion(completion: @escaping (Result<String?, Error>) -> Void) {
        let version = UIDevice.current.systemVersion
        completion(.success(version))
    }
    
    func getApps(completion: @escaping (Result<[UsedApp], Error>) -> Void) {
        let dummyApps = [
            UsedApp(id: "com.apple.mobilesafari", name: "Safari", minutesUsed: 120),
            UsedApp(id: "com.apple.mobilemail", name: "Mail", minutesUsed: 45),
            UsedApp(id: "com.apple.MobileSMS", name: "Messages", minutesUsed: 90),
            UsedApp(id: "com.apple.mobileslideshow", name: "Photos", minutesUsed: 30),
            UsedApp(id: "com.apple.mobilecal", name: "Calendar", minutesUsed: 15),
            UsedApp(id: "com.spotify.client", name: "Spotify", minutesUsed: 180),
            UsedApp(id: "com.netflix.Netflix", name: "Netflix", minutesUsed: 240),
            UsedApp(id: "com.instagram.ios", name: "Instagram", minutesUsed: 95),
            UsedApp(id: "com.facebook.Facebook", name: "Facebook", minutesUsed: 60),
            UsedApp(id: "com.twitter.ios", name: "Twitter", minutesUsed: 75)
        ]
        completion(.success(dummyApps))
    }
    
    func setAppTimeLimit(appId: String, minutesUsed: Int64, completion: @escaping (Result<StateResult, Error>) -> Void) {
        // Simulate different responses based on appId
        let appName = getAppName(from: appId)
        let result = StateResult(
            state: .success,
            message: "Time limit of \(minutesUsed) minutes set successfully for \(appName)"
        )
        completion(.success(result))
    }
    
    private func getAppName(from appId: String) -> String {
        switch appId {
        case "com.apple.mobilesafari":
            return "Safari"
        case "com.apple.mobilemail":
            return "Mail"
        case "com.apple.MobileSMS":
            return "Messages"
        case "com.apple.mobileslideshow":
            return "Photos"
        case "com.apple.mobilecal":
            return "Calendar"
        case "com.spotify.client":
            return "Spotify"
        case "com.netflix.Netflix":
            return "Netflix"
        case "com.instagram.ios":
            return "Instagram"
        case "com.facebook.Facebook":
            return "Facebook"
        case "com.twitter.ios":
            return "Twitter"
        default:
            return "Unknown App"
        }
    }
}

File: android/app/src/main/kotlin/com/example/ios_playground/AppUsageImplementation.kt
package com.example.ios_playground

class AppUsageImplementation : AppUsageApi {

    override fun getPlatformVersion(callback: (Result<String?>) -> Unit) {
        // Android version (e.g., 14)
        callback(Result.success(android.os.Build.VERSION.RELEASE))
    }

    override fun getApps(callback: (Result<List<UsedApp>>) -> Unit) {
        val dummyApps = listOf(
            UsedApp("com.android.chrome", "Chrome", 120L),
            UsedApp("com.google.android.gm", "Gmail", 45L),
            UsedApp("com.google.android.apps.messaging", "Messages", 90L),
            UsedApp("com.google.android.apps.photos", "Photos", 30L),
            UsedApp("com.google.android.calendar", "Calendar", 15L),
            UsedApp("com.spotify.music", "Spotify", 180L),
            UsedApp("com.netflix.mediaclient", "Netflix", 240L),
            UsedApp("com.instagram.android", "Instagram", 95L),
            UsedApp("com.facebook.katana", "Facebook", 60L),
            UsedApp("com.twitter.android", "Twitter", 75L)
        )
        callback(Result.success(dummyApps))
    }

    override fun setAppTimeLimit(
        appId: String,
        minutesUsed: Long,
        callback: (Result<StateResult>) -> Unit
    ) {
        val appName = getAppName(appId)
        val result = StateResult(
            state = State.SUCCESS,
            message = "Time limit of $minutesUsed minutes set successfully for $appName"
        )
        callback(Result.success(result))
    }

    private fun getAppName(appId: String): String {
        return when (appId) {
            "com.android.chrome" -> "Chrome"
            "com.google.android.gm" -> "Gmail"
            "com.google.android.apps.messaging" -> "Messages"
            "com.google.android.apps.photos" -> "Photos"
            "com.google.android.calendar" -> "Calendar"
            "com.spotify.music" -> "Spotify"
            "com.netflix.mediaclient" -> "Netflix"
            "com.instagram.android" -> "Instagram"
            "com.facebook.katana" -> "Facebook"
            "com.twitter.android" -> "Twitter"
            else -> "Unknown App"
        }
    }
}